### PR TITLE
Adds instructions for local deployment with microK8s

### PR DESCRIPTION
**Pull request by pedroleaoc**

### CODE REVIEW

**Comment by pedroleaoc** [comment]
Thanks @claudiubelu for helping to write this tutorial.

**Comment by pedroleaoc** [comment]
cc @claudiubelu to add instructions on /etc/hosts files.

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+Run `juju status` on your terminal, grab the `Address` of the `Unit` called `legend-sdlc/0*` and point your browser to `http://<SDLC_IP>:7070/api/auth/authorize`; Click on `Authorize` and you should see the text `Authorized`. 
+
+Similarly, run `juju status` on your terminal, grab the `Address` of the `Unit` called `legend-studio/0*` and point your browser to `http://<STUDIO-IP>:8080`. Click on `Authorize`. 
+
+If the process was sucessful, you will be able to see the Legend Studio dashboard on `http://<STUDIO-IP>:8080/studio/-/setup`! 🎉
+
+
+## Destroy setup
+To remove all deployed Legend applications, remove any data and reset Juju to the state it was before Legend was deployed, you can destroy the controller (created during the bootstrapping process). **This is a non-reversible operation - all your data created in studio will be lost!**
+
```
[the remark this answers is not in the document]
Comment (claudiubelu) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:149]: Also worth mentioning:

After destroying the ``finos-legend`` model, it no longer exists, so we can't ``juju status`` while we're still active on it. We have to switch to a different model that is available: ``juju switch controller``.

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+```bash
+juju show-unit gitlab-integrator/0 | grep legend-gitlab-redirect-uris
+```
+  
+Go back to the "Applications" page on GitLab.com and edit the application you created. Replace the dummy URI with the URIs retrieved from the command above and save the application. 
+
+Run `watch --color juju status --color` to see the applications reacting to the configuration change. As a result of this change, your FINOS Legend deployment should complete,
+
+## Accessing the Legend Studio dashboard
+
```
[the remark this answers is not in the document]
Comment (maoo) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:128]: @claudiubelu  - I've done that, but accessing `http://legend-sdlc:7070/api/auth/authorize` or `http://legend-studio:8080` hangs, then times out. Any idea why?

[the remark this answers is not in the document]
Comment (claudiubelu) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:128]: > @claudiubelu - I've done that, but accessing `http://legend-sdlc:7070/api/auth/authorize` or `http://legend-studio:8080` hangs, then times out. Any idea why?

It's just ``http://legend-sdlc/api/auth/authorize``, not ``http://legend-sdlc:7070/api/auth/authorize``. Due to the nginx-integrator charm, the ports are also mapped. For example, for Studio the port used by the application is ``8080``, but we map the standard port ``80`` to ``8080`` for Studio. For SDLC, we map the standard port ``80`` to ``7070`` used by SDLC and so on.

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+### Authorize the GitLab user and application
+
+Run `juju status` on your terminal, grab the `Address` of the `Unit` called `legend-sdlc/0*` and point your browser to `http://<SDLC_IP>:7070/api/auth/authorize`; Click on `Authorize` and you should see the text `Authorized`. 
+
+Similarly, run `juju status` on your terminal, grab the `Address` of the `Unit` called `legend-studio/0*` and point your browser to `http://<STUDIO-IP>:8080`. Click on `Authorize`. 
+
+If the process was sucessful, you will be able to see the Legend Studio dashboard on `http://<STUDIO-IP>:8080/studio/-/setup`! 🎉
+
+
```
Comment (claudiubelu) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:146]: Add section:

```
## Updating Legend applications

You can update and upgrade a charm automatically to the latest version by running ``juju upgrade-charm app-name --channel=edge``. For example, you can upgrade the Legend Studio charm by running: ``juju refresh legend-studio --channel=edge``.

However, if you want want to use a different Container Image than the charm has been published with, you will have to remove the application first, and add it manually, and then add the necessary relationships. You can see the currently created relations by running ``juju status --relations``.

Example: Let's say that you want to redeploy the Legend Studio with the latest ``finos/legend-studio:2.6.0`` image. For this, we need first remove the ``legend-studio`` charm:

``
juju remove-application legend-studio
``

You can see the status by running ``juju status``. After it was removed, redeploy it again, and also specifying the new Container Image as a resource:

``
juju deploy finos-legend-studio-k8s legend-studio --channel=edge --resource studio-image=finos/legend-studio:2.6.0
``

Afterwards, we need to add the necessary relations:

``
juju relate legend-studio legend-db
juju relate legend-studio legend-engine
juju relate legend-studio legend-sdlc
juju relate legend-studio gitlab-integrator
juju relate legend-studio ingress-studio
``

```

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (maoo) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:128]: Thanks @claudiubelu . Still timing out on ` http://legend-sdlc/api/auth/authorize`

My hosts file seems to be correct:
```
~ » ping legend-sdlc
PING legend-sdlc (127.0.1.1): 56 data bytes
```

How can I further debug this?

[the remark this answers is not in the document]
Comment (claudiubelu) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:128]: Pinging won't work, the nginx-integrator doesn't handle ping requests.

Can you do a ``juju status``?

[the remark this answers is not in the document]
Comment (maoo) [reply INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:135]:
```
Every 2.0s: juju status --color                                                                                                                                                                                                            maomini.lan: Tue Feb  1 12:14:55 2022

Model         Controller               Cloud/Region        Version  SLA          Timestamp
finos-legend  finos-legend-controller  microk8s/localhost  2.9.23   unsupported  12:14:55+01:00

App                Version  Status  Scale  Charm                               Store     Channel  Rev  OS          Address         Message
gitlab-integrator           active      1  finos-legend-gitlab-integrator-k8s  charmhub  edge      26  kubernetes  10.152.183.3
ingress-engine              active      1  nginx-ingress-integrator            charmhub  stable    24  kubernetes  10.152.183.113
ingress-sdlc                active      1  nginx-ingress-integrator            charmhub  stable    24  kubernetes  10.152.183.93
ingress-studio              active      1  nginx-ingress-integrator            charmhub  stable    24  kubernetes  10.152.183.194
legend-db                   active      1  finos-legend-db-k8s                 charmhub  edge      13  kubernetes  10.152.183.74
legend-engine               active      1  finos-legend-engine-k8s             charmhub  edge      14  kubernetes  10.152.183.229
legend-sdlc                 active      1  finos-legend-sdlc-k8s               charmhub  edge      37  kubernetes  10.152.183.245
legend-studio               active      1  finos-legend-studio-k8s             charmhub  edge      15  kubernetes  10.152.183.123
mongodb                     active      1  mongodb-k8s                         charmhub  edge      11  kubernetes  10.152.183.198

Unit                  Workload  Agent  Address      Ports  Message
gitlab-integrator/0*  active    idle   10.1.254.83
ingress-engine/0*     active    idle   10.1.254.84         Ingress with service IP(s): 10.152.183.164
ingress-sdlc/0*       active    idle   10.1.254.85         Ingress with service IP(s): 10.152.183.186
ingress-studio/0*     active    idle   10.1.254.86         Ingress with service IP(s): 10.152.183.2
legend-db/0*          active    idle   10.1.254.87
legend-engine/0*      active    idle   10.1.254.88
legend-sdlc/0*        active    idle   10.1.254.89
legend-studio/0*      active    idle   10.1.254.90
mongodb/0*            active    idle   10.1.254.91
```

I also invited you to https://finos-lf.slack.com/ where we have a chat on FINOS Legend integration for Juju. TY!

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+```bash
+sudo snap install microk8s --classic
+sudo snap alias microk8s.kubectl kubectl
+source ~/.profile
+sudo usermod -a -G microk8s $USER
+sudo chown -f -R $USER ~/.kube
+newgrp microk8s
+```
+
+On Mac, assuming you have XCode, Python 3.9 (with brew, you can run `brew link --overwrite python@3.9`) and [HomeBrew](brew.sh) installed, you can simply run `brew install ubuntu/microk8s/microk8s`.
```
Comment (claudiubelu) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:22]: Should mention that on Mac ``multipass`` is also required.

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+- Click "Applications" on the left menu. 
+- Create a new application with the following information:
+  - name
+  - Check the `Confidential` checkbox 
+  - Enter the following Redirect URIs:
+      ```
+      http://legend-studio/studio/log.in/callback
+      http://legend-engine/callback
+      http://legend-sdlc/api/auth/callback
+      http://legend-sdlc/api/pac4j/login/callback
```
Comment (claudiubelu) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:109]: Update to:

```
http://finos-legend/engine/callback
http://finos-legend/api/auth/callback
http://finos-legend/api/pac4j/login/callback
http://finos-legend/studio/log.in/callback
```

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+``` bash
+juju config gitlab-integrator gitlab-client-id="<Application ID>" gitlab-client-secret="<Secret Id>"
+```
+
+#### 3. Update the GitLab application URIs
+
+The command below will retrieve the GitLab URIs from the GitLab Integrator charm  
+``` bash
+juju show-unit --app gitlab-integrator/0 | python3 -W ignore -c "import json, sys, yaml; doc = yaml.load(sys.stdin); urls = [json.loads(r['application-data']['legend-gitlab-redirect-uris']) for r in doc['gitlab-integrator/0']['relation-info']]; print('\n'.join([url for url_list in urls for url in url_list]))"
+```
```
Comment (claudiubelu) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:131]: Update to:

```
The command bellow will output the GitLab URIs expected to be used by Legend:

```bash
juju run-action gitlab-integrator/0 get-redirect-uris --wait
``
```

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+> - clear the browser's cache 
+> - [if you are using Firefox](https://support.mozilla.org/en-US/kb/enhanced-tracking-protection-firefox-desktop#w_how-to-tell-when-firefox-is-protecting-you), disable Enchanced Tracking Protection for the Studio and SDLC pages.
+
+In order to access the local FINOS Legend Application, the following lines should be added to the `/etc/hosts` file:
+
+``` bash
+127.0.1.1 legend-studio
+127.0.1.1 legend-engine
+127.0.1.1 legend-sdlc
+```
```
Comment (claudiubelu) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:154]: Update to:

```
In order to access the local FINOS Legend Application, the following line should be added to the ``/etc/hosts`` file:

``
127.0.1.1 finos-legend
``
```

And below:

```
Adding this line will allow you to access Legend directly in your browser through the user-friendly name. ``finos-legend`` is the ``external-hostname`` [configured for the Legend Applications](#Configure-a-hostname-for-your-Legend-Applications). If you've provided a different name, use that instead.
```

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+my_hostname="legend-host"
+juju config legend-engine external-hostname=$my_hostname
+juju config legend-sdlc external-hostname=$my_hostname
+juju config legend-studio external-hostname=$my_hostname
+```
+
+## Setup and Configure GitLab
+To run Legend, you need to either run a GitLab instance somewhere, or use GitLab.com; the type of installation really depends on user's requirements, there is a secion in `DEPLOY_GITLAB.md` that talks about that (TODO).
+
+If this is your first experience with Legend, we suggest you starting with GitLab.com as per the instructions below. If, instead, you're interested to test a Legend deployment with a local GitLab, please follow instructions on `DEPLOY_GITLAB.md`.
```
Comment (balbirthomas) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:108]:
```suggestion
If this is your first experience with Legend, we suggest you start with GitLab.com as per the instructions below. If, instead, you're interested to test a Legend deployment with a local GitLab, please follow instructions on `DEPLOY_GITLAB.md`.
```

File: INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+If you are using gitlab.com, follow the following three steps.
+
+#### 1. Create a GitLab Application
+
+1. Create an account on gitlab.com
+2. Create an application:
+- Login GitLab.com, click your profile picture on the right upper corner and click "Preferences".
```
Comment (balbirthomas) [inline INSTRUCTIONS/LOCAL_DEPLOYMENT/README.md:116]:
```suggestion
- Login to GitLab.com, click your profile picture on the right upper corner and click "Preferences". 
```

**Pull request merged**